compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -196,15 +196,18 @@ class ExpressionCodegen(
         if (state.isParamAssertionsDisabled)
             return
 
-        val isSyntheticOrBridge = irFunction.origin.isSynthetic ||
+        val notCallableFromJava = isInlineLambda ||
+                Visibilities.isPrivate(irFunction.visibility) ||
+                irFunction.origin.isSynthetic ||
+                (irFunction is IrConstructor && irFunction.parentAsClass.isAnonymousObject) ||
```

Comment:
As I understand there would be similar problem with redundant assertions on captured variables in other cases,e.g.:
```
fun foo() {
    val captured = "123"
    
    class Z(val withAssertion: String) {
           
        val x = captured
    }
    
    Z("12")
}
```
So maybe it's best to specify special (synthetic?) kind for captured variables and variables passed to super constructor and don't generate for such variable assertions

Reply:
The important part here is not the assertion itself, which is never triggered but still valid code, but how the inliner handles it. Specifically, when the captured variable is an inline lambda, the load of that lambda before the null-checking intrinsic causes the variable remapper to throw an exception. Since local named classes in inline functions are currently rejected by the frontend, only the anonymous object case really needs to be handled right now.

Reply:
Could you add TODO about possible further logic generalization for mentioned cases?

Reply:
done